CodeSignal/dockerfiles
# Auto build dev on docker

## Changes
### Functional
Now circleci will build updated dockerfiles and push it to dockerhub with dev tag. So if you need to update many images at once, you don't need to go to each repo and configure custom build rules.
Check should fail if an image fails to build. Look at [this commit](https://circleci.com/gh/CodeSignal/dockerfiles/16). I broke the acceptance-tests dockerfile on purpose.
Check should fail if pushing to dockerhub fails. Look at [this run](https://circleci.com/gh/CodeSignal/dockerfiles/17) where I don't have permission to push to elixir repo.

New workflow if you need to updated image
1. Update docker file(s)
2. Create a PR, cicrleci will automatically build and push the image(s) to dockerhub with `dev` tag
3. You can go to circleci to look at build log to check whether build succeeded
4. Deploy a develop coderunner pointing to the image(s) with `dev` tag
5. If an image has dependencies, after releasing a new version of the image, repeat the process again for its dependencies

## Testing
I made some changes in elixir, r, d and acceptance-test.
1. Go to Circleci
2. Look at the tests. There should be build logs for elixir, d, r and acceptance-tests.
3. Rerun  workflow.
4. Go to dockerhub. Should see codesignal/elixir, codesignal/r-env, codesignal/d-env and codesignal/acceptance-tests dev tag updated recently.

* I will revert the dummy changes I made before merging. 
* Right now the credential saved as environment variables `DOCKERHUB_USERNAME` and `DOCKERHUB_PASSWORD` is my login. Maybe we should create a codesignal account for that purpose.
